### ISSUE TITLE
RFC: Service Worker

### ISSUE THREAD

**Issue by arv** [rocicorp/replicache-sdk-js#226]
**STATUS: as of Dec'20, we are NOT doing this. At least for now. Hit too many implementation issues.**

# Background

> Replicache seeks to enable "near-instant" reads for client-side applications. We define "near-instant" as <20ms for "a few screenfuls of data" for common applications... 

See #201 for more about the above.

Multiple tabs share the data stored in the Indexed DB, but each tab have their own instance of the Replicache JS SDK and its Wasm module. We currently do not do much to coordinate these tabs and we have these potential problems:

- With multiple tabs, syncs can happen in parallel. This is wasteful from a perf pov, and introduces lots of complexity / opportunity for bugs.
- Having multiple tabs means that we are relying on idb's built-in locking semantics, which are slightly different from ones we want, which is error-prone
- Implementing subscriptions seems easier with a service worker approach because there's a single writer and a built-in way to message with the listener
- Service worker could help alleviate jank when scanning

## Worker

Using a web worker we can do more work off the main thread, preventing us from blocking the UI for more than sending/sharing the actual data between the main thread and the worker.

## Shared Worker

Using a shared worker allows multiple tabs to share a single Wasm module.
 - This allows us to not have to worry about changes happening behind our back.
 - Only ever have one sync happening at the same time
 - No need to coordinate between tabs

However shared workers are not available on all platforms we want to support.

## Service Worker

The main use case for service workers (SW) was to allow intercepting HTTP requests to allow web apps to work offline. However, a service worker can be used form many other things and it cover the shared worker scenario too because  multiple tabs share a worker with the same scope (scope is the path that the SW should intercept, ie `./replicache-sw/`).

# Implementation

We currently abstract away all the calls to the Repc Wasm module behind an `Invoker` interface (`REPMWasmInvoker`. We also used to have an `REPMHTTPInvoker` as a workaround for other issues before). All these calls are async today so it is pretty easy to replace this invoker with another invoker that marshals the call to the service worker.

### Main thread

```js
export class REPMWorkerInvoker {
  async invoke(dbName, rpc, args) {
      // Pseudo code... needs more work to have it return a promise with the result
      return this._worker.postMessage({dbName, rpc, args});
  }
}
```

### Worker

```js
const wasmInvoker = new REPMWasmInvoker();

self.onmessage = async e => {
  const {dbName, rpc, args} = e.data;
  const res = await wasmInvoker.invoke(dbName, rpc, args);
  postMessage(res);
};
```

### `scan`

`scan` is currently not using a "pure" request/response model for performance reasons. Instead it uses a receiver function as part of its request argument and this function is called for every scan item. Scan then returns "void". To make the above marshalling model work we need to special case `scan` to not do this at the worker RPC boundary.

# Problems

## Task vs Microtask

`postMessage` uses a task. Tasks are enqueued on the browser message queue which means that things like render and input events can happen before the task continue. Microtasks are called when the js call stack is unwound, within the same task. Tasks are an order of magnitude slower.

## More resource files

Now we have 3 files that needs to be available on our customers server (replicache.js, replicache.wasm and replicache-sw.js). We've had issues with bundlers not doing the right thing when there are multiple files.

## ES modules

Support for modules in workers is only available in Chrome. This means we need to compile to an iife... Not a big deal but makes developing a little bit harder because we need to use compiled sources.

# Ideas

## Transferables

Workers generally do their message passing by copying the data. However, they support something called transferables, which allows passing the ownership of a binary array buffer to a worker/main thread.

We could then have repc return an array buffer that is transferred to the main thread and then decoded to JSON and then decoded to JS on the main thread.

### CBOR

Why not skip JSON completely in the RPC. [CBOR](http://cbor.io/) seems like a fast powerful schema less binary encoding

## ~Marshal Functions~

**I don't think this is a good idea. It is too confusing not a being able to call any other pure functions.**

Marshal the function (by source code) to the worker for `mutate`, `query` and `subscribe`. These functions are supposed to be pure anyway so it should be safe to do this. The benefit here is that we can continue to use microtasks for `get` and `has` and we only need to marshal the return value of these functions.

## Shared Array Buffer

This is not widely available yet. It is only available in Chrome desktop and Firefox. Shared Array Buffers got disabled after [meltdown/spectre](https://developers.google.com/web/updates/2018/02/meltdown-spectre) but are enabled in Chrome again and in Firefox with some special headers.

I believe we can use [Atomics.waitAsync](https://www.chromestatus.com/features/6243382101803008) to get back to microtask performance

**Comment by phritz** [comment]
I admit to not fully understanding how the current multi-tab model works and when I try to wrap my head around it I find it challenging to reason about. If for no other reason than architectural simplification I like the idea of a single repc shared by all tabs through which all replicache operations flow. Most of what I'd like to ask is for more detail in several areas. 

As mentioned, I don't fully understand how the current system works wrt multiple tabs, and so I have a little trouble understanding the delta. If you wouldn't mind, could you make an explicit comparison "here's how x,y,z work right now; here's how x,y,z work with SW". A sentence or two for each item would helpful. Things that I'd find it helpful to understand
  - rpcs: seems like we use a new, slower mechanism (postmessage) but that there may be some mitigation?
  - scan: not clear to me what changes with the callback?
  - subscribe: can you say briefly how it works now vs how will be implemented with SW
  - sync: will we do anything different?
  - how changes to db from sync initiated by one tab are signaled to other tabs
  - how changes to db from mutations in one tab are signaled to others
  - [dispatch](https://github.com/rocicorp/repc/blob/d0042d987135a2b42acd09ee554a27388d3c387c/src/wasm.rs#L19) is currently the entrypoint. Is there any issue with calling this concurrently at the top level of the SW in the new world? 
  - different deployments: customers may have two tabs open, one with their production app and one with their dev/staging version of the app, where the dev and prod versions use different replicache databases. I think this just works now, so long as they are using different dbs. Does anything change with SW? There's for example no way for the dev app to get signals from the prod SW?
  - different versions in different tabs: I think something Bad could happen now if there are two tabs at different versions of bindings+repc using the same db, yes? Should we do anything to address? And then is there an analog in the SW world ie the js running in the application tab is talking to a SW that has a js and a repc at different version? 
  - i think the potential issues mentioned in the "So..." section of https://github.com/rocicorp/repc/issues/240#issuecomment-727648664 might get worse in the SW world. Writes can now starve across tabs instead of just within one, and new read txs will not open until there are zero tx ops executing in any tab. I can't say whether this is any worse than we have today because I forget the idb details that come into play in the non-SW world. I suspect it is not worth worrying about now, just reminding myself that as you mentioned we are shifting cross-tab idb locking to SW repc locking.

Some thoughts and questions that occurred to me while reading:
  - any repc state that is kept in memory now becomes shared state. For some state eg log level this seems fine, but other state might have to be updated. For example the way database connections work might need to change. Right now closing a db in one tab has no effect on the others. But if two tabs are using the same db connection in the SW and one tab closes the connection we don't want it closed for the other. Might need open to increment and close to decrement a counter. The ConnMap is the main thing I'm thinking of. Looking for others a quick read through the code reveals RPC_COUNTER (seems fine to share), SENDER (seems fine, I think), TRANSACTION_COUNTER (seems fine), and TxnMap (seems fine, it just gets bigger) are also kept in memory.
  - logging: when we call console log from within a service worker is it visible in the application page's console log?
  - security: any concerns? I don't understand the access control model so not even sure what to ask :(
  - spawning the SW: my ignorance is showing here but if the app in two different tabs starts at the same time there is only ever one SW spawned, yes? 
  - update: again my ignorance here, suppose the app updates to a new version of bindings+repc, I assume there's a nice way for the old SW to die off or be replaced with the new one.
  - our [wasm tests](https://github.com/rocicorp/repc/blob/master/tests/wasm.rs) presumably just work in the new world? 
  - index changes: createindex with same name and metadta is a nop, but dropping a non-existence index by name is an error. maybe it should be a nop? i can't convince myself there is an issue with indexes other than that (eg, the app updates in two tabs simultaneously) but mentioning it in case someone else can think fo something. 

I'm a little unclear on whether the Transferables and CBOR are things that you are proposing we do? 

I'm also curious about the state of the performance delta you noticed between echo and repc. 

And finally... thanks for writing this up. It's really helpful for me to have things laid out like this because I do not have a full picture of all the pieces.

**Comment by arv** [comment]
Lots of questions. GH issues are not the best for a threaded discussion but I'll try:

Little bit more background to set the stage:

The current idea is to have the wasm module run in a worker. At this point we keep almost all of the js code on the main thread but this could change in the future. The main thread and the worker can communicate with `postMessage`. Internally this is done by posting a task between the threads. In the future we can consider using shared memory using a SharedArrayBuffer but it is not widely supported at the moment. If we use shared memory we can use Atomics to implement mutexes etc to achieve a much faster way of getting the data between the threads. 

Another thing to remember is that the SW is per URL scope. Only identical apps (same URL or at least same path for the js) might end up with the same SW.

- rpcs: seems like we use a new, slower mechanism (postmessage) but that there may be some mitigation?
  - Without SharedArrayBuffer we need to use postMessage which copies the data between the threads. There is also the issue regarding tasks vs microtasks. (tasks are slower).
- scan: not clear to me what changes with the callback?
  - All of our RPCs except scan takes input data and returns output data. Scan takes input data and a callback and returns nothing. Instead the callback is called multiple times with parts of the output data.
- subscribe: can you say briefly how it works now vs how will be implemented with SW
  - Today we only check if the root hash changed and we blindly call onData even if the data didn't actually change.
sync: will we do anything different?
  - #63 has a lot of details
- how changes to db from sync initiated by one tab are signaled to other tabs
  - Today they all sync and the only thing preventing them from stepping on each other is that they share the IDB
  - with a SW the worker would postMessage all the tabs. We could use BroadcastChannel if it wasn't for Safari crap. The SW can iterate over all its current clients instead.
- how changes to db from mutations in one tab are signaled to others
  - Today we notify different tabs using localStorage change events
  - We would broadcast this to all tabs (this would be done using subscribe)
- dispatch is currently the entrypoint. Is there any issue with calling this concurrently at the top level of the SW in the new world?
  - I don't think so
- different deployments in different tabs: customers may have two tabs open, one with their production app and one with their dev/staging version of the app, where the dev and prod versions use different replicache databases. I think this just works now, so long as they are using different dbs. Does anything change with SW? There's for example no way for the dev app to get signals from the prod SW?
  - The SW is registered using an URL scope. If we want to use different wasm modules we need to register the SW using a different scope.
  - If they use different db names
  - When broadcasting things we need to be careful to not send it to the wrong client/tabs
- different versions in different tabs: I think something Bad could happen now if there are two tabs at different versions of bindings+repc using the same db, yes? Should we do anything to address? And then is there an analog in the SW world ie the js running in the application tab is talking to a SW that has a js and a repc at different version?
  - Good point. I think we can include the version number in the SW scope and make sure we filter out the tabs as needed when broadcasting.
  - You can force the tabs and SW to reload if the versions are incompatible but all of this requires coordination
- i think the potential issues mentioned in the "So..." section of rocicorp/repc#240 (comment) might get worse in the SW world. Writes can now starve across tabs instead of just within one, and new read txs will not open until there are zero tx ops executing in any tab. I can't say whether this is any worse than we have today because I forget the idb details that come into play in the non-SW world. I suspect it is not worth worrying about now, just reminding myself that as you mentioned we are shifting cross-tab idb locking to SW repc locking.
  - I believe it is worse in a way that if one tab can stalls a write transaction today it only impacts that tab but it in the future it will impact all the tabs.
- any repc state that is kept in memory now becomes shared state. For some state eg log level this seems fine, but other state might have to be updated. For example the way database connections work might need to change. Right now closing a db in one tab has no effect on the others. But if two tabs are using the same db connection in the SW and one tab closes the connection we don't want it closed for the other. Might need open to increment and close to decrement a counter. The ConnMap is the main thing I'm thinking of. Looking for others a quick read through the code reveals RPC_COUNTER (seems fine to share), SENDER (seems fine, I think), TRANSACTION_COUNTER (seems fine), and TxnMap (seems fine, it just gets bigger) are also kept in memory.
  - open/close is definitely an issue right now
- logging: when we call console log from within a service worker is it visible in the application page's console log?
  - Each tab and service worker have their own log. Devtools lets you switch between them but I think by default it includes the current tab and its associated service workers.
- security: any concerns? I don't understand the access control model so not even sure what to ask :(
  - SW uses the same original model
- spawning the SW: my ignorance is showing here but if the app in two different tabs starts at the same time there is only ever one SW spawned, yes?
  -Correct. The SW is associated with the URL of the SW js file as well as the scope (URL "pattern").
- update: again my ignorance here, suppose the app updates to a new version of bindings+repc, I assume there's a nice way for the old SW to die off or be replaced with the new one.
  - Yes. The browser fetches the SW js file and if it changes it initiates an update. If you do nothing the SW will not be reloaded until all tabs are closed but you can force the SW to be reloaded even when there are tabs using it.
- our wasm tests presumably just work in the new world?
  - Yes. They operate on a different layer
- index changes: createindex with same name and metadata is a nop, but dropping a non-existent index by name is an error. maybe it should be a nop? i can't convince myself there is an issue with indexes other than that (eg, the app updates in two tabs simultaneously) but mentioning it in case someone else can think fo something.
  - I think in general if two tabs try to migrate the data at the same time there will be issues
- I'm a little unclear on whether the Transferables and CBOR are things that you are proposing we do?
  - I think it is worth doing performance tests using binary data all the way (from main thread to Rust and back to main thread). JSON and UTF8 encoding/decoding is something we do too much already and adding another layer means more of those.
- I'm also curious about the state of the performance delta you noticed between echo and repc.
  - More about this later

**Comment by aboodman** [comment]
On Wed, Dec 2, 2020 at 9:50 AM Erik Arvidsson <notifications@github.com>
wrote:

> Lots of questions. GH issues are not the best for a threaded discussion
> but I'll try:
>
> Little bit more background to set the stage:
>
> The current idea is to have the wasm module run in a worker. At this point
> we keep almost all of the js code on the main thread but this could change
> in the future. The main thread and the worker can communicate with
> postMessage. Internally this is done by posting a task between the
> threads. In the future we can consider using shared memory using a
> SharedArrayBuffer but it is not widely supported at the moment. If we use
> shared memory we can use Atomics to implement mutexes etc to achieve a much
> faster way of getting the data between the threads.
>
> Another thing to remember is that the SW is per URL scope. Only identical
> apps (same URL or at least same path for the js) might end up with the same
> SW.
>
>    - rpcs: seems like we use a new, slower mechanism (postmessage) but
>    that there may be some mitigation?
>       - Without SharedArrayBuffer we need to use postMessage which copies
>       the data between the threads. There is also the issue regarding tasks vs
>       microtasks. (tasks are slower).
>    - scan: not clear to me what changes with the callback?
>       - All of our RPCs except scan takes input data and returns output
>       data. Scan takes input data and a callback and returns nothing. Instead the
>       callback is called multiple times with parts of the output data.
>    - subscribe: can you say briefly how it works now vs how will be
>    implemented with SW
>       - Today we only check if the root hash changed and we blindly call
>       onData even if the data didn't actually change.
>       sync: will we do anything different?
>       - #63 <https://github.com/rocicorp/replicache-sdk-js/issues/63> has
>       a lot of details
>    - how changes to db from sync initiated by one tab are signaled to
>    other tabs
>       - Today they all sync and the only thing preventing them from
>       stepping on each other is that they share the IDB
>       - with a SW the worker would postMessage all the tabs. We could use
>       BroadcastChannel if it wasn't for Safari crap. The SW can iterate over all
>       its current clients instead.
>    - how changes to db from mutations in one tab are signaled to others
>       - Today we notify different tabs using localStorage change events
>       - We would broadcast this to all tabs (this would be done using
>       subscribe)
>    - dispatch is currently the entrypoint. Is there any issue with
>    calling this concurrently at the top level of the SW in the new world?
>       - I don't think so
>    - different deployments in different tabs: customers may have two tabs
>    open, one with their production app and one with their dev/staging version
>    of the app, where the dev and prod versions use different replicache
>    databases. I think this just works now, so long as they are using different
>    dbs. Does anything change with SW? There's for example no way for the dev
>    app to get signals from the prod SW?
>       - The SW is registered using an URL scope. If we want to use
>       different wasm modules we need to register the SW using a different scope.
>       - If they use different db names
>       - When broadcasting things we need to be careful to not send it to
>       the wrong client/tabs
>    - different versions in different tabs: I think something Bad could
>    happen now if there are two tabs at different versions of bindings+repc
>    using the same db, yes? Should we do anything to address? And then is there
>    an analog in the SW world ie the js running in the application tab is
>    talking to a SW that has a js and a repc at different version?
>       - Good point. I think we can include the version number in the SW
>       scope and make sure we filter out the tabs as needed when broadcasting.
>       - You can force the tabs and SW to reload if the versions are
>       incompatible but all of this requires coordination
>
> The sort of default path that Service Worker APIs lead you down prevent
this from happening. It's not possible for the service worker to upgrade
until all the tabs referencing it close. There are ways to be more clever
but I do not think we would want to use them.

>
>    -
>       - i think the potential issues mentioned in the "So..." section of
>    rocicorp/repc#240 <https://github.com/rocicorp/repc/issues/240>
>    (comment) might get worse in the SW world. Writes can now starve across
>    tabs instead of just within one, and new read txs will not open until there
>    are zero tx ops executing in any tab. I can't say whether this is any worse
>    than we have today because I forget the idb details that come into play in
>    the non-SW world. I suspect it is not worth worrying about now, just
>    reminding myself that as you mentioned we are shifting cross-tab idb
>    locking to SW repc locking.
>       - I believe it is worse in a way that if one tab can stalls a write
>       transaction today it only impacts that tab but it in the future it will
>       impact all the tabs.
>    - any repc state that is kept in memory now becomes shared state. For
>    some state eg log level this seems fine, but other state might have to be
>    updated. For example the way database connections work might need to
>    change. Right now closing a db in one tab has no effect on the others. But
>    if two tabs are using the same db connection in the SW and one tab closes
>    the connection we don't want it closed for the other. Might need open to
>    increment and close to decrement a counter. The ConnMap is the main thing
>    I'm thinking of. Looking for others a quick read through the code reveals
>    RPC_COUNTER (seems fine to share), SENDER (seems fine, I think),
>    TRANSACTION_COUNTER (seems fine), and TxnMap (seems fine, it just gets
>    bigger) are also kept in memory.
>       - open/close is definitely an issue right now
>
> Hm. This is actually quite interesting. Currently we can only have *one
connection* per database per repc instance. This is enforced in do_open.
Duplicate calls to open effectively share the same connection. We do all
our locking on a per-connection basis, so we don't want to change this.

It's worth asking what even is a "connection" in repc. It's not a lot. It's
the unit we implement transactionality at, so in that sense it is critical.
But otherwise, the only functionality it really has is the ability to
listen to changes and to get closed.

What does "close()" do? Is it basically a way to wait for all open tx to
stop and then prevent further tx from happening? What I'm getting at is
that maybe close doesn't make sense or isn't needed in a multitab world.

>
>    -
>       - logging: when we call console log from within a service worker is
>    it visible in the application page's console log?
>       - Each tab and service worker have their own log. Devtools lets you
>       switch between them but I think by default it includes the current tab and
>       its associated service workers.
>    - security: any concerns? I don't understand the access control model
>    so not even sure what to ask :(
>       - SW uses the same original model
>    - spawning the SW: my ignorance is showing here but if the app in two
>    different tabs starts at the same time there is only ever one SW spawned,
>    yes?
>    -Correct. The SW is associated with the URL of the SW js file as well
>    as the scope (URL "pattern").
>    - update: again my ignorance here, suppose the app updates to a new
>    version of bindings+repc, I assume there's a nice way for the old SW to die
>    off or be replaced with the new one.
>       - Yes. The browser fetches the SW js file and if it changes it
>       initiates an update. If you do nothing the SW will not be reloaded until
>       all tabs are closed but you can force the SW to be reloaded even when there
>       are tabs using it.
>    - our wasm tests presumably just work in the new world?
>       - Yes. They operate on a different layer
>    - index changes: createindex with same name and metadata is a nop, but
>    dropping a non-existent index by name is an error. maybe it should be a
>    nop? i can't convince myself there is an issue with indexes other than that
>    (eg, the app updates in two tabs simultaneously) but mentioning it in case
>    someone else can think fo something.
>       - I think in general if two tabs try to migrate the data at the
>       same time there will be issues
>
> I think we either need the index APIs to have read versions (getIndex,
hasIndex) and be transactional, or else have them be idempotent (so
dropIndex("non-existent") should not throw).

**Comment by phritz** [comment]
Thanks for writing this up @arv. This all makes sense. 

Regarding version upgrades (a tab moving from replicache v N to N+1), version skew (two tabs at different versions N and M), and deployment env differences (a tab with the production version of the app and a tab with a dev version of the app) I think we should not do any work in these departments rn but that we should note either now how they do currently work ("you can encode a version in your url" or whatever) or that we need to do some work to enable them and have a super brief associated issue. Maybe a Details doc linked to from the readme? Or server setup? Many people won't care but some will and it is frustrating if you care not being able to find mention of the issues in docs.

Callback: think we talked in zoom about working out a batching strategy, would love to see it here and happy to help figure out what makes sense if you would like. 

> What does "close()" do? Is it basically a way to wait for all open tx to
> stop and then prevent further tx from happening? What I'm getting at is
> that maybe close doesn't make sense or isn't needed in a multitab world.

It does not seem strictly necessary but does seem nice to have. You can be sure if there are zero open connections to a db that you can do whatever you want to it, safely delete it, migrate it, whatever. In tests  the ability to close a db ensures there's no state left over to affect other tests. So while I agree with you, it does seem "clean" to have it for some vague notion of cleanliness.

>  I think we either need the index APIs to have read versions (getIndex,
> hasIndex) and be transactional, or else have them be idempotent (so
> dropIndex("non-existent") should not throw).

Agree that drop should not throw if the db does not exist.

**Comment by arv** [comment]
## Scan

Here is one possible approach

1. `main`: Create a `MessagePort`. This is going to be used exclusively for the scan operation
2. `main`: send a `scan` RPC to the worker. Also send one endpoint of the `MessagePort` over.
3. `worker`: Invoke the Wasm RPC with a callback that accumulates results
  1. When we have accumulated X results, post those back on the dedicated port.
4. `main`: Listen to messages on the dedicated port and feed those into the `ScanIterator`
5. `worker`: When the Wasm RPC is done return the result to the main thread (the result is going to be null/void)
6. `main`: When main worker RPC is done the scan results are done too.

We do not really need two different `MessageChannel`s here and we can see what makes sense here to make the code cleaner.

I also believe we should fix #230 before starting this part.

**Comment by aboodman** [comment]
Close for now.
